src/Polynomial3_TEST.cc

```diff
@@ -0,0 +1,153 @@
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+*/
+#include <gtest/gtest.h>
+#include <ostream>
+
+#include "ignition/math/Polynomial3.hh"
+
+using namespace ignition;
+
+/////////////////////////////////////////////////
+TEST(Polynomial3Test, DefaultConstructor)
+{
+  const math::Polynomial3d poly;
+  EXPECT_EQ(poly.Coeffs(), math::Vector4d::Zero);
+}
+
+/////////////////////////////////////////////////
+TEST(Polynomial3Test, Constructor)
+{
+  const math::Polynomial3d poly(math::Vector4d::One);
+  EXPECT_EQ(poly.Coeffs(), math::Vector4d::One);
+}
+
+/////////////////////////////////////////////////
+TEST(Polynomial3Test, ConstructionHelpers)
+{
+  const math::Polynomial3d poly = math::Polynomial3d::Constant(1.);
+  EXPECT_EQ(poly.Coeffs(), math::Vector4d(0., 0., 0., 1.));
+}
+
+/////////////////////////////////////////////////
+TEST(Polynomial3Test, Evaluate)
+{
+  {
+    const math::Polynomial3d p =
+        math::Polynomial3d::Constant(1.);
+    EXPECT_DOUBLE_EQ(p(-1.), 1.);
+    EXPECT_DOUBLE_EQ(p(0.), 1.);
+    EXPECT_DOUBLE_EQ(p(1.), 1.);
+    EXPECT_DOUBLE_EQ(p(math::INF_D), 1.);
+  }
+  {
+    const math::Polynomial3d p(math::Vector4d::One);
+    EXPECT_DOUBLE_EQ(p(-1.), 0.);
+    EXPECT_DOUBLE_EQ(p(0.), 1.);
+    EXPECT_DOUBLE_EQ(p(1.), 4.);
+    EXPECT_DOUBLE_EQ(p(-math::INF_D), -math::INF_D);
+  }
+}
+
+/////////////////////////////////////////////////
+TEST(Polynomial3Test, Minimum)
+{
+  {
+    const math::Polynomial3d p0 =
+        math::Polynomial3d::Constant(1.);
+    EXPECT_TRUE(std::isnan(p0.Minimum(
+        math::Intervald::Open(0., 0.))));
+    EXPECT_DOUBLE_EQ(p0.Minimum(), 1.);
+  }
+  {
+    const math::Polynomial3d p1(
+        math::Vector4d(0., 0., 1., 1.));
+    EXPECT_TRUE(std::isnan(p1.Minimum(
+        math::Intervald::Open(0., 0.))));
+    EXPECT_DOUBLE_EQ(p1.Minimum(
+        math::Intervald::Open(0., 1.)), 1.);
+    EXPECT_DOUBLE_EQ(p1.Minimum(), -math::INF_D);
+  }
+  {
+    const math::Polynomial3d p2(
+        math::Vector4d(0., 1., 1., 1.));
+    EXPECT_TRUE(std::isnan(p2.Minimum(
+        math::Intervald::Open(0., 0.))));
+    EXPECT_DOUBLE_EQ(p2.Minimum(
+        math::Intervald::Open(1., 2.)), 3.);
+    EXPECT_DOUBLE_EQ(p2.Minimum(), 0.75);
+  }
+  {
+    const math::Polynomial3d p3(
+        math::Vector4d(1., 1., 1., 1.));
+    EXPECT_TRUE(std::isnan(p3.Minimum(
+        math::Intervald::Open(0., 0.))));
+    EXPECT_DOUBLE_EQ(p3.Minimum(
+        math::Intervald::Open(-1., 1.)), 0.);
+    EXPECT_DOUBLE_EQ(p3.Minimum(), -math::INF_D);
+  }
+  {
+    const math::Polynomial3d p4(
+        math::Vector4d(-1., 2., 1., 1.));
+    EXPECT_TRUE(std::isnan(p4.Minimum(
+        math::Intervald::Open(0., 0.))));
+    EXPECT_DOUBLE_EQ(p4.Minimum(
+        math::Intervald::Open(-1., 1.)), 3);
```

Comment:
I'm not sure this is correct. I'm graphing this on this site:

https://www.mathopenref.com/cubicexplorer.html

![image](https://user-images.githubusercontent.com/5751272/159547127-ea5ee4a9-4ed0-48e5-b7b6-b01f2e36af81.png)

In the `(-1, 1)` interval, the minimum seems to be `< 3`, at that local minimum. Am I missing something here?

For example, `p4.Evaluate(-0.3) == 0.907 < 3`

Reply:
Oof, I had a brainfart. Fixed in https://github.com/ignitionrobotics/ign-math/pull/393/commits/0108ccfa1bcfe77c92ea6d1c68d30e61d554b491. I was missing a few parenthesis.

Reply:
Thanks! I'm reviewing it again now. One thing I forgot to mention yesterday is that I think it would be helpful for users of this function if it also returned the X at minimum, not just Y. Is it too much to ask that this be added? I could be a separate function, or the signature of the current function could be changed slightly, i.e.

```
Vector2<T> Minimum(const Interval<T> &_interval) const
```

or 

```
void Minimum(const Interval<T> &_interval, T &_x, T &_y) const
```

or something else

Reply:
Thanks for the fix in the logic, I can't find a way to break it :smile: While reviewing I added some more test expectations to make sure different parts of the curve were handled properly, as well as curves with different characteristics. Do you mind adding them here? This should help with branch coverage.

<details><summary>More expectations</summary>
<p>

```diff
diff --git a/src/Polynomial3_TEST.cc b/src/Polynomial3_TEST.cc
index 842d232d..50e9e334 100644
--- a/src/Polynomial3_TEST.cc
+++ b/src/Polynomial3_TEST.cc
@@ -99,6 +99,10 @@ TEST(Polynomial3Test, Minimum)
         math::Intervald::Open(0., 0.))));
     EXPECT_DOUBLE_EQ(p.Minimum(
         math::Intervald::Open(1., 2.)), 3.);
+    EXPECT_DOUBLE_EQ(p.Minimum(
+        math::Intervald::Open(-1., 0.)), 0.75);
+    EXPECT_DOUBLE_EQ(p.Minimum(
+        math::Intervald::Open(-3., -2.)), 3);
     EXPECT_DOUBLE_EQ(p.Minimum(), 0.75);
   }
   {
@@ -108,6 +112,10 @@ TEST(Polynomial3Test, Minimum)
         math::Intervald::Open(0., 0.))));
     EXPECT_DOUBLE_EQ(p.Minimum(
         math::Intervald::Open(1., 2.)), -1.);
+    EXPECT_DOUBLE_EQ(p.Minimum(
+        math::Intervald::Open(-2., -1.)), -5.);
+    EXPECT_DOUBLE_EQ(p.Minimum(
+        math::Intervald::Open(0., 1.)), 1.);
     EXPECT_DOUBLE_EQ(p.Minimum(), -math::INF_D);
   }
   {
@@ -117,6 +125,10 @@ TEST(Polynomial3Test, Minimum)
         math::Intervald::Open(0., 0.))));
     EXPECT_DOUBLE_EQ(p.Minimum(
         math::Intervald::Open(-1., 1.)), 0.);
+    EXPECT_DOUBLE_EQ(p.Minimum(
+        math::Intervald::Open(-2., -1.)), -5.);
+    EXPECT_DOUBLE_EQ(p.Minimum(
+        math::Intervald::Open(2., 3.)), 15.);
     EXPECT_DOUBLE_EQ(p.Minimum(), -math::INF_D);
   }
   {
@@ -125,8 +137,13 @@ TEST(Polynomial3Test, Minimum)
     EXPECT_TRUE(std::isnan(p.Minimum(
         math::Intervald::Open(0., 0.))));
     EXPECT_DOUBLE_EQ(p.Minimum(
-        math::Intervald::Open(-1., 1.)),
-        0.8873882090776197);
+        math::Intervald::Open(-1., 1.)), 0.8873882090776197);
+    EXPECT_DOUBLE_EQ(p.Minimum(
+        math::Intervald::Open(-3., -2.)), 15);
+    EXPECT_DOUBLE_EQ(p.Minimum(
+        math::Intervald::Open(1., 2.)), 3);
+    EXPECT_DOUBLE_EQ(p.Minimum(
+        math::Intervald::Open(2., 3.)), -5);
     EXPECT_DOUBLE_EQ(p.Minimum(), -math::INF_D);
   }
   {
@@ -136,6 +153,19 @@ TEST(Polynomial3Test, Minimum)
         math::Intervald::Open(0., 0.))));
     EXPECT_DOUBLE_EQ(p.Minimum(
         math::Intervald::Open(-1., 1.)), -3);
+    EXPECT_DOUBLE_EQ(p.Minimum(
+        math::Intervald::Open(0., 2.)), 1);
+    EXPECT_DOUBLE_EQ(p.Minimum(
+        math::Intervald::Open(2., 3.)), 3);
+    EXPECT_DOUBLE_EQ(p.Minimum(), -math::INF_D);
+  }
+  {
+    const math::Polynomial3d p(
+        math::Vector4d(1., -4., -2., -1.));
+    EXPECT_TRUE(std::isnan(p.Minimum(
+        math::Intervald::Open(0., 0.))));
+    EXPECT_DOUBLE_EQ(p.Minimum(
+        math::Intervald::Open(-1., 6.)), -16.051047904897441);
     EXPECT_DOUBLE_EQ(p.Minimum(), -math::INF_D);
   }
 }
```

</p>
</details>

Reply:
Done and done in https://github.com/ignitionrobotics/ign-math/pull/393/commits/3f4fadca83426fc3d71cedd0dd3fb4bf79721fae and https://github.com/ignitionrobotics/ign-math/pull/393/commits/79d493c879e93cc8f6c12a3baeb80a77dc731e9b, respectively